### PR TITLE
add source_file_url and source_line to spec

### PR DESCRIPTION
**Pull request by live1206**
[AB#196431](https://dev.azure.com/ceapex/Engineering/_workitems/edit/196431)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/5770)

### CODE REVIEW

File: docs/designs/link-output.md [hunk 1]
```diff
@@ -57,20 +60,26 @@ Example:
         {
             "source_url": "/dotnet/articles/a",
             "source_moniker_group": "218c13d0",
-            "target_url": "/dotnet/articles/b"
+            "target_url": "/dotnet/articles/b",
+            "source_file_url": "https://github.com/dotnet/articles/a.md",
```
Comment (herohua) [inline docs/designs/link-output.md:64]: if the content is from CRR repo, would `source_file_url` be pointing to that repo?

Comment (live1206) [reply docs/designs/link-output.md:64]: Theoretically yes, it should a URL pointing to the CRR repo. Will add a spec test for this scenario.

File: docs/designs/link-output.md [hunk 2]
```diff
@@ -57,20 +60,26 @@ Example:
...
-            "target_url": "./c.md"
+            "target_url": "./c.md",
+            "source_file_url": "https://github.com/dotnet/articles/a.md",
+            "source_line": "30"
         }
     ]
 }
 ```
-> Notice: `source_url` should be the root file for inclusion (TOC inclusion and markdown inclusion)
+> Notice: `source_url` and `source_file_url` should be the root file for inclusion (TOC inclusion and markdown inclusion)
```
Comment (OsmondJiang) [inline docs/designs/link-output.md:84]: is it possible to point to inclusion file's github url for the links belongs to inclusion?

File: docs/designs/link-output.md [hunk 3]
```diff
@@ -42,11 +42,14 @@ Docfx generate `links` information of each files into one single file `.links.js
...
-  | `source_url`            | /dotnet/articles/a   |
-  | `source_moniker_group`  | 218c13d0             |
-  | `target_url`            | /dotnet/articles/b   |
+  |                         |                                         |
+  |-------------------------|----------------------                   |
+  | `source_url`            | /dotnet/articles/a                      |
+  | `source_moniker_group`  | 218c13d0                                |
+  | `target_url`            | /dotnet/articles/b                      |
+  | `source_file_url`       | https://github.com/dotnet/articles/a.md |
+  | `source_line`           | 10                                      |
```
Comment (herohua) [inline docs/designs/link-output.md:51]: `source_line` will be null if not applicable

Comment (live1206) [reply docs/designs/link-output.md:51]: Planning to set 1 if null

File: docs/designs/link-output.md (hunk 2 above)
Comment (live1206) [reply docs/designs/link-output.md:84]: you are right, will update the note

File: docs/designs/link-output.md (hunk 3 above)
Comment (herohua) [reply docs/designs/link-output.md:51]: that's okay to me - but would be better we call it out as well as the index is one-based

Comment (superyyrrzz) [inline docs/designs/link-output.md:51]: Say `a.md` includes `b.md`. Will `source_line` be the line of `[!include]` in `a.md`, or the line of actual broken link in `b.md`?

Comment (OsmondJiang) [reply docs/designs/link-output.md:51]: the source file url should point to b.md and source line should be the line in b.md, that's my understanding.

Comment (live1206) [reply docs/designs/link-output.md:51]: the `source_line` is from `SourceInfo`, the comments says it is [one based start](https://github.com/dotnet/docfx/blob/805251fa713f08ac58894762c6673c64713268b7/src/docfx/lib/log/SourceInfo.cs#L18)

Comment (live1206) [reply docs/designs/link-output.md:51]: > the source file url should point to b.md and source line should be the line in b.md, that's my understanding.

agree

File: docs/designs/link-output.md [hunk 4]
```diff
@@ -57,20 +60,26 @@ Example:
...
-            "target_url": "./c.md"
+            "target_url": "./c.md",
+            "source_git_url": "https://github.com/dotnet/articles/a.md",
+            "source_line": "30"
         }
     ]
 }
 ```
-> Notice: `source_url` should be the root file for inclusion (TOC inclusion and markdown inclusion)
+> Notice: `source_url` should be the root file for inclusion (TOC inclusion and markdown inclusion), and `source_file_url` should be the included file
```
Comment (herohua) [inline docs/designs/link-output.md:84]: would this cause too much work to v2 @superyyrrzz

Comment (superyyrrzz) [reply docs/designs/link-output.md:84]: It is OK for v2.

The issue for v2 is it is hard to report line number under current implementation. If this is resolved, it is easier to report the line number from included file than from the including file.

**Pull request merged**